bindings/python/iree/runtime/tracing.py

```diff
@@ -0,0 +1,164 @@
+"""Tracing support."""
+
+# Copyright 2021 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from genericpath import exists
+from typing import Dict, List, Optional, Sequence
+
+import logging
+import os
+import sys
+
+from . import binding as _binding
+
+try:
+  import yaml
+except ModuleNotFoundError:
+  _has_yaml = False
+else:
+  _has_yaml = True
+
+__all__ = [
+    "get_default_tracer",
+    "Tracer",
+    "TRACE_PATH_ENV_KEY",
+]
+
+TRACE_PATH_ENV_KEY = "IREE_SAVE_CALLS"
+
+
+class Tracer:
+  """Object for tracing calls made into the runtime."""
+
+  def __init__(self, trace_path: str):
+    if not _has_yaml:
+      self.enabled = False
+      logging.warning("PyYAML not installed: tracing will be disabled")
+      return
+    self.enabled = True
+    self.trace_path = trace_path
+    os.makedirs(trace_path, exist_ok=True)
+    self._name_count = dict()  # type: Dict[str, int]
+
+  def persist_vm_module(self, vm_module: _binding.VmModule) -> "TracedModule":
+    # Depending on how the module was created, there are different bits
+    # of information available to reconstruct.
+    name = vm_module.name
+    flatbuffer_blob = vm_module.stashed_flatbuffer_blob
+    if flatbuffer_blob:
+      save_path = os.path.join(self.trace_path,
+                               self.get_unique_name(f"{name}.vmfb"))
+      logging.info("Saving traced vmfb to %s", save_path)
+      with open(save_path, "wb") as f:
+        f.write(flatbuffer_blob)
+      return TracedModule(self, vm_module, save_path)
+
+    # No persistent form, but likely they are built-in modules.
+    return TracedModule(self, vm_module)
+
+  def get_unique_name(self, local_name: str) -> str:
+    if local_name not in self._name_count:
+      self._name_count[local_name] = 1
+      return local_name
+    stem, ext = os.path.splitext(local_name)
+    index = self._name_count[local_name]
+    self._name_count[local_name] += 1
+    unique_name = f"{stem}__{index}{ext}"
```

Comment:
(out of curiosity, why 2 underscores? a convention?)

Reply:
Because I'm not exhaustively deduping names and this was a simple way to avoid the common case where someone uses a single underscore in their module name.